kotlinx-coroutines-core/common/test/sync/MutexTest.kt

```diff
@@ -59,6 +59,21 @@ class MutexTest : TestBase() {
         assertFalse(mutex.isLocked)
     }
 
+    @Test
+    fun testWithLockFailureUnlocksTheMutex() = runTest {
+        val mutex = Mutex()
+        assertFalse(mutex.isLocked)
+        try {
+            mutex.withLock {
+                expect(1)
+                throw TestException()
+            }
+        } catch (e: TestException) {
+            finish(2)
+        }
+        assertFalse(mutex.isLocked)
+    }
```

Comment:
Tested more thoroughly in https://github.com/Kotlin/kotlinx.coroutines/pull/3986